### PR TITLE
Add cudf::strings::reverse function

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Adds `cudf::strings::reverse` function.
This is to support https://github.com/NVIDIA/spark-rapids/issues/6885

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/tests/strings/reverse_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+#include <cudf/strings/strings_column_view.hpp>
+
+#include <vector>
+
+struct StringsReverseTest : public cudf::test::BaseFixture {
+};
+
+TEST_F(StringsReverseTest, Reverse)
+{
+  cudf::test::strings_column_wrapper input({"abcdef", "12345", "", "", "aébé", "A é Z"});
```
Comment (bdice) [inline cpp/tests/strings/reverse_tests.cpp:32]: Can we include a null value in some test?

Comment (davidwendt) [reply cpp/tests/strings/reverse_tests.cpp:32]: Yes, I have more tests coming. This is a draft PR.

File: cpp/src/strings/reverse.cu [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+}  // namespace
+
+std::unique_ptr<column> reverse(strings_column_view const& input,
+                                rmm::cuda_stream_view stream,
+                                rmm::mr::device_memory_resource* mr)
+{
+  if (input.is_empty()) { return make_empty_column(type_id::STRING); }
+
+  // copy the column; replace data in the chars column
+  auto result = std::make_unique<column>(input.parent(), stream, mr);
```
Comment (hyperbolic2346) [inline cpp/src/strings/reverse.cu:65]: Is this making a full copy of the string data and then replacing it in the functor?

Comment (davidwendt) [reply cpp/src/strings/reverse.cu:65]: Yes, that was the intention of the comment. Should've been worded better?

**Review by robertmaynard — APPROVED** [review]
Approving CMakes changes

File: cpp/benchmarks/string/reverse.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  auto const num_rows  = static_cast<cudf::size_type>(state.get_int64("num_rows"));
+  auto const row_width = static_cast<cudf::size_type>(state.get_int64("row_width"));
+
+  if (static_cast<std::size_t>(num_rows) * static_cast<std::size_t>(row_width) >=
+      static_cast<std::size_t>(std::numeric_limits<cudf::size_type>::max())) {
+    state.skip("Skip benchmarks greater than size_type limit");
+  }
+
+  data_profile const table_profile = data_profile_builder().distribution(
+    cudf::type_id::STRING, distribution_id::NORMAL, 0, row_width);
```
Comment (ttnghia) [inline cpp/benchmarks/string/reverse.cpp:38]: I'm not sure if generating benchmark data this way will have 2-bytes characters?

File: cpp/src/strings/reverse.cu (hunk 2 above)
Comment (ttnghia) [reply cpp/src/strings/reverse.cu:65]: I see that the char child of the result column will be overwritten. Should we just copy the offsets over?

File: cpp/src/strings/reverse.cu [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  column_device_view const d_strings;
+  offset_type const* d_offsets;
+  char* d_chars;
+
+  __device__ void operator()(size_type idx)
+  {
+    if (d_strings.is_null(idx)) { return; }
+    auto const d_str = d_strings.element<string_view>(idx);
+    // pointer to the end of the output string
+    auto d_output = d_chars + d_offsets[idx] + d_str.size_bytes();
```
Comment (ttnghia) [inline cpp/src/strings/reverse.cu:48]: Is this `d_offsets` always zero-based? There is a test with sliced column below so I think it should be 0-based and just want to ask to confirm.

File: cpp/src/strings/reverse.cu (hunk 2 above)
Comment (davidwendt) [reply cpp/src/strings/reverse.cu:65]: This one step does everything we need. It builds the output column and normalizes the offsets if the input is sliced.

File: cpp/benchmarks/string/reverse.cpp (hunk 3 above)
Comment (davidwendt) [reply cpp/benchmarks/string/reverse.cpp:38]: The `create_random_table` does generate multibyte UTF-8 characters for `STRING` columns.

File: cpp/src/strings/reverse.cu (hunk 4 above)
Comment (davidwendt) [reply cpp/src/strings/reverse.cu:48]: The copy-column step ensures the offsets here will always be 0-based here.

**Review by ttnghia — APPROVED** [review]

File: cpp/tests/strings/reverse_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+struct StringsReverseTest : public cudf::test::BaseFixture {
+};
+
+TEST_F(StringsReverseTest, Reverse)
+{
+  auto input = cudf::test::strings_column_wrapper(
+    {"abcdef", "12345", "", "", "aébé", "A é Z", "X", "é"}, {1, 1, 1, 0, 1, 1, 1, 1});
+  auto results  = cudf::strings::reverse(cudf::strings_column_view(input));
+  auto expected = cudf::test::strings_column_wrapper(
+    {"fedcba", "54321", "", "", "ébéa", "Z é A", "X", "é"}, {1, 1, 1, 0, 1, 1, 1, 1});
```
Comment (cwharris) [inline cpp/tests/strings/reverse_tests.cpp:37]: Are any of these multiple byte characters? Should we have multiple byte characters in the test? I don't recall if we support them or not.

**Review by cwharris — APPROVED** [review]
really nice!

File: cpp/tests/strings/reverse_tests.cpp (hunk 5 above)
Comment (davidwendt) [reply cpp/tests/strings/reverse_tests.cpp:37]: All of the accented ones (`é`) are multi-byte UTF-8 characters.

**Review by AyodeAwe — APPROVED** [review]

**Review by bdice — APPROVED** [review]
LGTM.

One interesting thing that came up while thinking hard about reversed strings: [Emoji ZWJ sequences](https://emojipedia.org/emoji-zwj-sequence/). I don't think we need special handling here, just wanted to share my findings for curiosity.

An emoji like [Woman Shrugging, Medium Skin Tone](https://emojipedia.org/woman-shrugging-medium-skin-tone/) 🤷🏽‍♀️ is composed of a sequence of 5 code points:
[🤷 U+1F937](https://emojipedia.org/emoji/%F0%9F%A4%B7/)
[🏽 U+1F3FD](https://emojipedia.org/emoji/%F0%9F%8F%BD/)
U+200D
[♀ U+2640](https://emojipedia.org/emoji/%E2%99%80/)
U+FE0F

Python reverses the sequence literally, with the code points in the opposite order, thereby no longer making a single emoji but a string that renders like `️♀‍🏽🤷`. I would expect this PR to  behave similarly because each Unicode code point is handled individually, so I don't think additional testing or work is needed.

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**